### ISSUE TITLE
Use placeholder selectors when declaring classes

### ISSUE THREAD

**Issue by architarious** [wvuweb/wvu-design-system#124]
## IDEA ##
[Placeholder selectors](https://sass-lang.com/documentation/style-rules/placeholder-selectors) allow you to extend classes without creating inheritance problems or code bloat. Not only are they the officially suggested way of extending classes in a SASS library, but they'll be crucial in regards to reusing and creating new components going forward.

## PROBLEM ##
Someone want's to extend class#1 inside of class#2, but instead of just extending class#1 it extends class#2 everywhere that class#1 is being used in the entire stylesheet. Mixins can help to mitigate this, but they can lead to massive code bloat over time because they're essentially rewriting the same chunk of code over and over.

### Example ### 
```scss
.foo{ color: blue; }

.footer .foo{ font-weight: bold; }

.bar{ @extend .foo; }
```
### Outputs ###
```css
.foo, .bar{ color: blue; }

.footer .foo, .footer .bar{ font-weight: bold; }
```

## SOLUTION ##
When declaring a class, take the extra second to declare a silent class so that it can be extended elsewhere without breaking things. 

### Example ###
```scss
.foo, %foo{ color: blue; }

.footer .foo{ font-weight: bold; }

.bar { @extend %foo; }
```

### Outputs ###
```css
.foo, .bar{ color: blue; }

.footer .foo{ font-weight: bold; }
```

**Comment by adamjohnson** [comment]
What files would you want to use placeholder selectors with?

Also: https://www.sitepoint.com/avoid-sass-extend/

> For such a small example, the difference is quite minimal however it can quickly ramp up with larger content for the mixin/placeholder. So again, on paper it makes sense to think that extending placeholders rather than including mixins is better for file size.

> However files are usually served gzipped from the server. Gzip is based on DEFLATE and LZ77; quickly put, the more a string is repeated, the better it is for compression. When putting this in context, it means that the difference is likely to be inexistant once Gzip has rolled over the whole thing.

> This is exactly the same debate as for duplicated media queries when putting them inside selectors. It makes no difference once Gzip has done its thing:

>> … we hashed out whether there were performance implications of combining vs scattering Media Queries and came to the conclusion that the difference, while ugly, is minimal at worst, essentially non-existent at best.
>> Sam Richards in regards to Breakpoint

> So, using @extend rather than mixins because of code duplication is not a valid point. Moving on.

**Comment by architarious** [comment]
Ideally, I'd prefer if all of our branded components followed a structure similar to this:

```scss
@mixin shout( 
  $textColor: $wvublue, 
  $fontSize: $text-large){ 
    color: $textColor; 
    font-size: $fontSize;
    float:left;
    ....
} .shout, %shout{ @include(shout); }
```
That way you get to choose whether or not using a mixin or extending is the best fit depending on the scenario. I don't know if that's too tedious of an example to expect others to follow or not though; I will say that following that format definitely saves me a LOT of time in the long-run, especially when I need to make new updates or create new templates.

As far as performance goes, gzipping definitely shrinks the [differences between the two](https://gist.github.com/HugoGiraudel/9488917#gistcomment-1206190), but as long as your extending placeholders rather than classes you shouldn't have to worry about massive chunks of selectors being scattered everywhere. As long as your not trying to extend micro-classes like "%bold-sans-serif" or "%flexbox-row", this really shouldn't be a problem. 

Also, keep in mind that when we're discussing gziping we're only talking about transfer rates over a network, the browser still has to parse the code once it receives it. This is where extending comes in handy because it effectively [minifies](https://css-tricks.com/the-difference-between-minification-and-gzipping/) your code and [improves render time](http://carlos.bueno.org/2010/02/measuring-javascript-parse-and-load.html#minifi) for end users. Basically, it's always just going to be a lot quicker for a browser or anything else to parse `.a, .b { float:left}` than `.a{float:left} .b{float:left;}` because there's technically fewer lines and characters; the only way that changes is by trying to do it with everything.

In short, one isn't inherently "better" than the other, they have different use-cases. Mixins are great whenever your dealing with variables or are using the same small chunk of code everywhere because it helps with compression. Extending placeholders of mid/upper level components is great because it's  generally done sparingly and because it makes for minified code.

**Comment by adamglenn** [comment]
There are no classes like that though (.wvu-hero). The only classes we have are utilities. Would it help if I added them to the utility classes? It is no problem, only takes a few minutes. I was doing it, but at some point I just forgot to go back and add them in.

**Comment by architarious** [comment]
I was thinking there would eventually be a few more component classes, but as it is right now, i'd just make sure that stuff like slash and wvu-calendar have it because they may run into inheritance issues depending on how they're being used. I don't know, if you'd need it on all utitlities, but it's definetly something to consider if there are any that are over 10-15 lines of code.

**Comment by adamglenn** [comment]
I can see these being a useful alternative if folks would rather style their components in the way you suggested. Not everyone will want to use strictly utilities. I think what we'll probably do is if someone wants to do that, they can submit their version of the component and we can include it on the Design System site, along with a Sass partial that they can just quickly grab and incorporate into their own project. When we do this, we'll be sure to include the % extend code. Or if they want to grab a component that extends another, we can include instructions in the implementation notes.
